homeassistant/components/media_source/local_source.py

```diff
@@ -0,0 +1,167 @@
+from .models import BrowseMedia, MediaSource, MediaSourceItem, PlayMedia
+
+
+@callback
+def async_setup(hass: HomeAssistant):
+    """Set up local media source."""
+    source = LocalSource(hass)
+    hass.data[DOMAIN][DOMAIN] = source
+    hass.http.register_view(LocalMediaView(hass))
+
+
+class LocalSource(MediaSource):
+    """Provide local directories as media sources."""
+
+    def __init__(self, hass: HomeAssistant):
+        """Initialize local source."""
+        super().__init__(DOMAIN)
+        self.hass = hass
+
+    @callback
+    def async_parse_identifier(self, item: MediaSourceItem):
+        """Parse identifier."""
+        if item.identifier == "":
+            source_dir_id = "media"
+            location = ""
+
+        else:
+            source_dir_id, location = item.identifier[
+                item.identifier.startswith("/") and 1 :
+            ].split("/", 1)
+
+        if source_dir_id != "media":
+            raise Unresolvable("Unknown source directory.")
+
+        if location != sanitize_path(location):
+            raise Unresolvable("Invalid path.")
+
+        return source_dir_id, location
+
+    @callback
+    def async_full_path(self, source_dir_id, location) -> Path:
+        """Return full path."""
+        return self.hass.config.path("media", location)
+
+    async def async_resolve_media(self, item: MediaSourceItem) -> str:
+        """Resolve media to a url."""
+        source_dir_id, location = self.async_parse_identifier(item)
+        mime_type, _ = await self.hass.async_add_executor_job(
+            mimetypes.guess_type, self.async_full_path(source_dir_id, location)
+        )
+        return PlayMedia(item.identifier, mime_type)
+
+    async def async_browse_media(
+        self, item: MediaSourceItem, media_types: Tuple[str] = MEDIA_MIME_TYPES
+    ) -> BrowseMedia:
+        """Return media."""
+        try:
+            source_dir_id, location = self.async_parse_identifier(item)
+        except Unresolvable as err:
+            raise BrowseError(str(err))
+
+        return await self.hass.async_add_executor_job(
+            self._browse_media, source_dir_id, location
+        )
+
+    def _browse_media(self, source_dir_id, location):
+        """Browse media."""
+        full_path = Path(self.hass.config.path("media", location))
+
+        if not full_path.exists():
+            raise BrowseError("Path does not exist in.")
+
+        if not full_path.is_dir():
+            raise BrowseError("Path is not a directory.")
+
+        return self._build_item_response(source_dir_id, full_path)
+
+    def _build_item_response(self, source_dir_id: str, path: Path, is_child=False):
+        mime_type, _ = mimetypes.guess_type(str(path))
+        media = BrowseMedia(
+            DOMAIN,
+            f"{source_dir_id}/{path.relative_to(self.hass.config.path('media'))}",
+            path.name,
+            mime_type,
+        )
+        media.can_play = path.is_file()
```

Comment:
This is redundant, if not even wrong since `BrowseMedia` expects `can_play: bool` as fourth argument.
```suggestion
        )
        media.can_play = path.is_file()
```